### DISCUSSION TITLE
[Feature request](MS6): Server Pack Exports

### DISCUSSION THREAD

**Discussion by kreezxil** [Griefed/ServerPackCreator#589] [Ideas]
### Whats the feature you would like to be added?

Have SPC create a mod manifest, much in the same way that the official pack launchers make their exports. This would allow for a few things, the main part of which will be a drastically smaller server pack (pre-installed) size to upload to CurseForge.

### Anything else you would like to add?

Currently, most of the FTB mod packs, under the server files section, has a binary that is distributed that acts like a client, it says it's gathering the files and eventually shows itself downloading them.

However, and I don't remember where, this was not a binary, it was a script that did it.

I think you're already scanning the cf API anyway, meaning you're probably not that far off from creating this already.

It would be useful for the super mega packs, which sadly will die off because people would rather not play mod packs that go above a gigabyte anyway. Not until the poverty class computer is a 32 GB standard computer, at least.

for reference: https://www.feed-the-beast.com/modpack/91_ftb_oceanblock/server-files

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Griefed** [comment]
Heya Kreezxil,

thanks for this feature request. 🙂

I have indeed already thought about adding such a feature. Namely, adding an option to export a [ServerStarter](https://github.com/BloodyMods/ServerStarter) compatible format.

> However, and I don't remember where, this was not a binary, it was a script that did it.

Correct me if I am wrong. This sounds a lot like: https://github.com/BloodyMods/ServerStarter
I have no idea how the FTB people generate the OS binaries which you linked below, but I would prefer OS-independant exports anyway. Generate once, run anywhere. Plus, if their installers download from CurseForge, we run into the CurseForge API and token problem again. See below.

> I think you're already scanning the cf API anyway, meaning you're probably not that far off from creating this already.

Unfortunately, due to the changes/limitations to the CurseForge API, I am not. At least not currently.
A rewrite of the way SeverPackCreator accesses CurseForges API is still planned, but it will require the user to set a valid CurseForge API token in the configuration in order to access CurseForge related features. Otherwise it will create either of the following situations:
1. Some checks/values are gathered from CurseForge until a rate limit is hit, thus effectively killing the feature for any potential user and generating incomplete server packs, i.e garbage.
2. A given user, without a valid CurseForge API token, can not generate such packages/exports etc..

It is not possible to ship a CurseForge API token with ServerPackCreator *without* compromising it. ServerPackCreator is open source, so anyone can figure out how to get a hold of a token, should any be embedded into it. Or they could use a decompiler and get it this way.

If and when the API rewrite for SPC is done, I may consider this as a feature for the webservice. There, the admin(s) who run it can set a token and any generated server pack could have the option to additionally include a ServerStarter export.

Of course, tests need to be made whether ServerStarter still works after the CurseForge API changes. It may be entirely possible that ServerStarter has become useless due to these changes. So, preferrably, ServerStarter tests *must* happen *before* I start working on this. If ServerStarter has become unreliable due to CurseForges changes, this feature is pretty much already dead, unfortunately.

I don't see any other feasible way of reliably implementing this.

TL;DR: Exports would be awesome, but thanks to CurseForges API changes, they would only be available on a webservice instance with a valid token configured. ServerStarter functionality must be evaluated first, then I can come back to this when the CurseForge API implementation is done.

Cheers,
Griefed

**Comment by Griefed** [comment]
CurseForge/Modrinth API implementation has been scrapped completely.

I have yet to come up with a good idea how server pack exports could be realized in a useful way and whether this feature would make sense for ServerPackCreator. Personally, I would put this under "Out of scope", but oh well.

Modrinth manifests provide a lot of information about a modpack, and ServerPackCreator is going to make use of them. Maybe someone will come up with a good idea sometime, but currently, that someone is not me. 😂

**Comment by kreezxil** [comment]
No pressure, just leave it on the back burner, close it for now. but maybe
pin it too, so someone can see it's been requested. That way someone
willing to figure out the alchemy required and which type of goat the
demons require for a sacrifice can then make that part happen. Maybe add a
tag to it such as "good idea for a plugin".

**Comment by Griefed** [comment]
Alrighty, I'll move it to the discussions-section.

**Comment by Griefed** [comment]
See discussion: https://github.com/Griefed/ServerPackCreator/discussions/345

Cheers,
Griefed

**Comment by Griefed** [comment]
Version 6, the webservice rework, will support this feature.
Currently, the plan is to have the webservice exclusively use the Modrinth API to create server pack scripts of a given Modrinth modpack which, upon execution, will then download all required mods and setup the modloader etc.

**Comment by Griefed** [comment]
Heyo,

long time no speaky.
As you may have noticed, V6 didn't get the export functionality I mentioned in a previous comment.

I've decided against using the Modrinth API for various reasons. The main reason being that CurseForge modpacks are nigh-on impossible to "convert" or "translate" to modrinth packs in terms of the mods which are used. Either mods aren't on Modrinth or under a different name or other version etc. pp. 

Haven't had a good idea for server pack exports so far.

I did find a nice article which explains how to stuff files and folders into a selfextracting shell-script and added a section to the help-website, but this is not an official part of the ServerPackCreator app due to the reasons listed in the help-article, and it only works on Linux-systems: https://help.serverpackcreator.de/fun-stuff.html#fun-stuff

**Comment by kreezxil** [comment]
looking at my opening statement i don't see where I suggested that cf mod packs be made into server packs using modrinth mods and vice versa. To me, the optimal solution is that modrinth packs generate modrinth server packs, and cf packs generate cf server packs. That's the easiest way.

sadly that ocean is heavily muddied by pack devs insisting on making their packs live on both platforms, which is crazy. other content i can understand but packs no.

**Comment by Griefed** [comment]
It was just an example. The other way around, Modrinth to CurseForge is just as nigh-on impossible to "convert" or "translate". Either way, some platform is required which hosts the files contained in the server pack, but the diversity in the modpacks makes this incredibly difficult to realize from my end of things.

**Comment by Griefed** [comment]
Moved the issue to the discussions, as there is not concrete idea on how this request could be achieved.
